### PR TITLE
Support authentication using gcp workload identity federation

### PR DESCRIPTION
**Pull request by akhil-rane**
Workload identity federation allows authentication with GCP using short-lived
token-based credentials. The current version of golang.org/x/oauth2 does not 
support it. Upgrading the version.

This PR also includes changes that will allow gcp storage driver code to consume
new `external account` type credentials that support workload identity.

ref: https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity

### CODE REVIEW

**Comment by akhil-rane** [comment]
@milosgajdos This PR adds a new authentication method for gcs storage driver to authenticate with GCP. Please take a look.

**Comment by milosgajdos** [comment]
Hey, @akhil-rane. Busy with $DAY_JOB, I'm afraid.

Furthermore, unfortunately, I'm not familiar with GCP Cloud stack well enough to make my review worth anything other than: "This seems ok?". If I find some proper time for this on Friday I'll look more closely. In the meantime I'll defer to the other maintainers.

**Comment by akhil-rane** [comment]
Thanks for @milosgajdos Do we want to tag someone who is familiar with GCP driver code here?

**Comment by milosgajdos** [comment]
I've added Gitlab folks -- they _might_ be familiar with GCP stack

**Comment by deleteriousEffect** [comment]
> @joaodrp @deleteriousEffect Did you get a chance to take a look at this?

@akhil-rane We haven't, sorry. Although we work a bit with GCP we don't do a lot of the admin side, so we don't have special knowledge about identity federation. 

I'll put aside some time next week to test these changes locally. In general, I think they look good.

File: registry/storage/driver/gcs/gcs_test.go
Comment (deleteriousEffect) [inline registry/storage/driver/gcs/gcs_test.go:77]: This isn't a bad change, but this is out of scope for this PR. This should be handled in a separate PR so that we can better track it in git history.

**Review by deleteriousEffect — CHANGES_REQUESTED** [review]

**Comment by deleteriousEffect** [comment]
@akhil-rane could you please target [distribution:release/2.8](https://github.com/distribution/distribution/tree/release/2.8) instead? That's the latest 2.x branch now.

**Comment by akhil-rane** [comment]
> @akhil-rane could you please target [distribution:release/2.8](https://github.com/distribution/distribution/tree/release/2.8) instead? That's the latest 2.x branch now

done

**Comment by DavidSpek** [comment]
This is similar to https://github.com/distribution/distribution/pull/3921 and it's been decided these types of changes won't be merged into the current release branch anymore and instead all effort is being put into getting the v3 release out.

@akhil-rane Could you check if a change is needed on the main branch to be able to use workload identity? If so, this PR should instead target the main branch or a new PR should be opened to address the issue there.

**Comment by milosgajdos** [comment]
With https://github.com/distribution/distribution/releases/tag/v3.0.0-alpha.1 out now, I'm inclined to close this PR as it won't be merged into `2.x` branch. All the further development is going into getting a stable `v3` release out next year.

@akhil-rane note that there has been a major overhaul of the GCS driver in https://github.com/distribution/distribution/pull/4120 which have addressed the issues your PR attempts to address.

**Comment by milosgajdos** [comment]
Closing (see my last comment), feel free to reopen @akhil-rane

**Pull request closed without merging**
